### ISSUE TITLE
Dart Native to SQLite C direct?

### ISSUE THREAD

**Issue by stevesmithworks** [tekartik/sqflite#200]
I found the following article written by @sma about interfacing with the SQLite C API directly from Dart.

[Creating an SQLite package for Dart](https://medium.com/@eibaan_54644/creating-an-sqlite-package-for-dart-4c27ffe43ee0)

Would this approach of using Dart directly to C have better performance than the current sqflite approach of using Objective-C for iOS and Java for Android? Would the C API be something sqflite could evolve to use?

**Comment by sma** [comment]
I can't comment on the performance of sqflite. But I'm not sure whether creating your own shared native library is even possible with Flutter. As far as I know the only documented way is using the kind of plugins sqflite is and using method channels to communicate with the Flutter engine.

I noticed that sqflite uses FMDB instead of directly accessing sqlite3 via C. But I don't think that this has negative impact on the performance. My guess would be that serializing and deserializing of data transfered by method channels would takes the most time. The only way to omit this would be to stay in Dart-land and build your own database engine using Dart. As the file format is documented (https://www.sqlite.org/fileformat.html) one could theoretically access an sqlite3 db file directly with Dart – at least for reading stuff. But doing so seems to be a lot of work…

**Comment by alextekartik** [comment]
Nice promising post indeed. Waiting for it to allow targetting mobile and desktop

**Comment by venkatd** [comment]
Another relevant article
https://github.com/dart-lang/sdk/blob/master/samples/ffi/sqlite/docs/sqlite-tutorial.md

**Comment by jaumard** [comment]
FFI as now land into flutter stable :) still not 100% optimized but the benchmark around crypto show it's already faster than full dart solution. Curious to see sqlite using this to get rid of plugin need ^^

**Comment by alextekartik** [comment]
I see several issues still.

* Database connection

On Android, if you exit (back) an app and restarts it right away, the native process is not killed but the dart app itself is killed and restarted (not sure I use the right wording here). This lead to several issues where there might still be an open connection to a database and if a transaction was in progress, there is no way you can re-open a connection. Sqflite handles this on the native side by trying to reuse an existing connection and eventually cancelling (rollback) any pending transaction)

If we use ffi, I’m not sure how this could be handled in a safe way; in the best case, an app will need to make sure every connection is closed before exiting.

* Threading

iOS and Android seems to differ in their implementation. Sqflite starts a thread for each database, this ensures the proper behavior of transaction. With ffi if we call sqlite directly, a long operation might simply hang the UI as it is in the main thread. Maybe isolate will be a solution but that still means some messaging needed between the main isolate and an sqlite isolate. If we use native thread, well, that means using ffi for that too or writing a native part to handle threading like the current plugin does, this native part could be really different for each platform.

* Binary distribution

For now sqflite uses the platform implementation. Some people might want to keep it this way, some want a new well defined sqlite implementation, which will mean more binary code to bundle and it is not clear how this will work and how this could target mobile and desktop.

* So...

This might not be as easy as it sounds.

SQLite seems a good target for ffi and I think I get an email almost everyday asking when sqflite will support that. For now I don’t have the skills nor dedicated time to tackle such a task. I know the moor package is doing some experiments.

I’d rather see first a good sqlite native library first (i.e. pushing the current experimentation to a real implementation that works in a real application!). Once that is done, I’ll be happy to write a wrapper to provide the same api sqlite has for compatibility. The API  has some drawbacks (no cursor) and advantages (callback for migration). If you use it well, sqflite can handle huge databases and complex queries efficiently. If you use it badly, ffi won’t solve much.

So yes, I’m following that. I have not tried ffi myself yet but if someone else could handle this not so easy task, that would be great!

**Comment by stevesmithworks** [comment]
The highest possible performance SQLite integration in a Flutter app is so critical for app development that I hope the @flutter team will take on the development of a native implementation.

**Comment by lamkr** [comment]
I am developing a dart package that uses the native sqlite3 library. If you would like to contribute or give suggestions, please see https://github.com/lamkr/sqlite3dart

**Comment by jayjah** [comment]
Actually I just want to post the following youtube video, because it clears a lot stuff which was mentioned here:
[Youtube | Flutter - C interoperability with Dart FFI | Session](https://www.youtube.com/watch?v=2MMK7YoFgaA&ab_channel=Flutter)

Spoiler:
They show a working example of using sqlite with dart ffi, which supports in the end the following platforms: 
ios, android, macOs, Linux and Windows.

**Comment by dcharkes** [comment]
We're working on FFI plugin support that should fully support this use case:

* https://github.com/flutter/flutter/pull/96225

> Would this approach of using Dart directly to C have better performance than the current sqflite approach of using Objective-C for iOS and Java for Android?

Most certainly. sqlite is a native library (written in C/C++). So when going through Java or Swift, it will end up in C at end but with multiple jumps. Directly calling to C/C++ with Dart FFI should be faster. If not, please ping me with reproductions so that I can triage these.

> With ffi if we call sqlite directly, a long operation might simply hang the UI as it is in the main thread.

I would suggest using an extra isolate to do these calls.

> Maybe isolate will be a solution but that still means some messaging needed between the main isolate and an sqlite isolate.

These messages should be faster than putting them through method channels (especially with the new lightweight isolates).

> For now sqflite uses the platform implementation.

Correct, that is indeed something we do not have a solution for yet when using Dart FFI.

> Binary distribution

You could also opt to distribute the source in a third_party folder and build it during Flutter application builds.

Distributing pre-built binaries in FFI plugins is future work:

* https://github.com/flutter/flutter/issues/96465

**Comment by alextekartik** [comment]
@dcharkes Good thing to have better support for ffi. However why make it flutter only? `sqflite_common_ffi` works on the DartVM (Linux, MacOS, android) and flutter too and runs the query in a separate isolate (and use the `sqlite3` ffi package). When used with `sqlite_flutter_libs` it also works on iOS and Android

**Comment by dcharkes** [comment]
> When used with `sqlite_flutter_libs` it also works on iOS and Android

Ah, I was under the impression that it did not wort through FFI in Flutter apps because of:

> * See sqflite for flutter mobile
> * See sqflite_common_ffi for Desktop and unit test

https://github.com/tekartik/sqflite/blob/master/sqflite_common/README.md 

Cool that it already works! 🚀 🎸 

> However why make it flutter only?

The goal is to support Flutter & Dart Standalone in one package in the future. But we can't do everything in a single PR. For Flutter we can use `flutter ...` to do native builds. But for Dart standalone we have no standardized way yet. You can see some of our vision in https://flutter.dev/go/ffi-plugins.

**Comment by alextekartik** [comment]
https://pub.dev/packages/sqflite_common_ffi through sqlite3 does exactly that!
